manager/controllers/default/security/login.class.php

```diff
@@ -281,20 +286,105 @@ public function handleForgotLoginHash()
+     * @return void
+     */
+    public function handleMagicLoginLink() {
+
+        // Handle new password form
+        if (!empty($_GET['magiclink'])) {
+            $hash = $this->modx->sanitizeString($_GET['magiclink']);
+            /** @var modDbRegister $registry */
+            $registry = $this->modx->getService('registry', 'registry.modRegistry')
+                ->getRegister('user', 'registry.modDbRegister');
+            $registry->connect();
+            $registry->subscribe('/pwd/change/' . $hash);
+
+            $record = $registry->read(['poll_limit' => 1, 'remove_read' => false]);
+
+//            if (!empty($registry->read(['poll_limit' => 1, 'remove_read' => false]))) {
+//                $this->scriptProperties['magiclink'] = $this->modx->sanitizeString($hash);
+//                // Reassign lexicons to smarty so we could use system setting here
+//                $this->placeholders['_lang']['login_new_password_note'] = $this->modx->lexicon('login_new_password_note', [
+//                    'length' =>$this->modx->getOption('password_min_length')
+//                ]);
+//                $this->modx->smarty->assign('_lang', $this->placeholders['_lang']);
+            $this->modx->log(modX::LOG_LEVEL_ERROR, "LOG ME MAGICALLY IN");
+
+            /** @var modUser $user */
+            if (empty($record) || !$user = $this->modx->getObject('modUser', ['username' => reset($record)])) {
+                $this->modx->smarty->assign('error_message', $this->modx->lexicon('login_magiclink_err'));
+
+                return;
+            }
+
+            /** @var modUserProfile $profile */
+            $profile = $user->getOne('Profile');
+            $this->scriptProperties['passwordgenmethod'] = 's';
+
+            $this->scriptProperties['passwordnotifymethod'] = 'no';
+            $this->scriptProperties['newPassword'] = true;
+            $this->modx->lexicon->load('core:user');
+
+            if (!class_exists('modUserUpdateProcessor')) {
+                require(MODX_CORE_PATH . 'model/modx/processors/security/user/update.class.php');
+            }
+            $processor = new modUserUpdateProcessor($this->modx, $this->scriptProperties);
+            $processor->modx->error->reset();
+            if (!class_exists('modUserValidation')) {
+                require(MODX_CORE_PATH . 'model/modx/processors/security/user/_validation.php');
+            }
+
+
+            $password = uniqid("tmp-password-");
+            $user->set('password', $password);
+            $user->save();
+
+            $this->scriptProperties['username'] = $user->get('username');
+            $this->scriptProperties['password'] = $password;
+            $registry->read(['poll_limit' => 1, 'remove_read' => true]);
+
+            // TODO remove magiclink
+            $this->setActivationHash('', $user);
+
+            /** @var modProcessorResponse $response */
+            $response = $this->modx->runProcessor('security/login', $this->scriptProperties);
```

Comment:
I'll need to review this more closely, but I'm a little confused we need another security/login call in this controller. It seems to be duplicating fairly complex logic that's also in the handleLogin method and would be best to not repeat.

Reply:
Right, its actually the code copied from the existent login stuff. Going to refactor that.